File: ova/10004/bicolor.py
import sys

def main():
    vertices = int(input())
    if vertices == 0:
        sys.exit(0)
    nedges = int(input())

    edges = {}
    for _ in range(nedges):
        x,y = map(int, input().split())
        try:
            edges[x].append(y)
        except KeyError:
            edges[x] = [y]

    reds = set([0])
    blacks = set()
    paint_black = True
    queue = [0]
    while True:
        v = queue.pop()
        if v in edges.keys():
            if paint_black:
                for v2 in edges[v]:
                    if v2 in reds:
                        print("NOT BICOLORABLE.")
                        return
                    blacks.add(v2)
                    if v2 not in queue:
                        queue.append(v2)
                    paint_black = False
            else:
                for v2 in edges[v]:
                    if v2 in blacks:
                        print("NOT BICOLORABLE.")
                        return
                    reds.add(v2)
                    if v2 not in queue:
                        queue.append(v2)
                    paint_black = True
        if queue == []:
            break
    print("BICOLORABLE.")


if __name__ == "__main__":
    while True:
        main()

File: ova/10004/bicolor.cpp
#include <iostream>
#include <list>
#include <vector>
#include <set>
#include <queue>

using namespace std;

int main()
{
    int edges, lines;
    int n1, n2;
    bool paint_black, bicolorable;


    while(true) {
        cin >> edges;
        if(edges == 0) return 0;
        vector<list<int> > G(edges);
        set<int> reds, blacks;
        queue<int> q;

        cin >> lines;
        for(int i = 0; i < lines; ++i) {
            cin >> n1 >> n2;
            G[n1].push_back(n2);
            G[n2].push_back(n1);
        }

        bicolorable = true;
        reds.insert(0);
        q.push(0);
        while(!q.empty() and bicolorable) {
            auto node = q.front();
            q.pop();
            paint_black = reds.find(node) != reds.end();
            for(auto it = G[node].begin(); it != G[node].end(); ++it) {
                if(paint_black) {
                    if(reds.find(*it) != reds.end()) {
                        cout << "NOT BICOLORABLE." << endl;
                        bicolorable = false;
                        break;
                    } else {
                        auto new_elem = blacks.insert(*it);
                        if(new_elem.second == true) q.push(*it); 
                    }
                } else {
                     if(blacks.find(*it) != blacks.end()) {
                        cout << "NOT BICOLORABLE." << endl;
                        bicolorable = false;
                        break;
                    } else {
                        auto new_elem = reds.insert(*it);
                        if(new_elem.second == true) q.push(*it); 
                    }
                }
            }
        }
        if(bicolorable) cout << "BICOLORABLE." << endl;
    }

    return 0;
}
